src/engine.rs

```diff
@@ -1,41 +1,24 @@
 use std::io::BufRead;
 use std::sync::atomic::{AtomicIsize, AtomicUsize, Ordering};
-use std::sync::{Arc, RwLock};
+use std::sync::{Arc, Mutex, RwLock};
 use std::time::Instant;
-use std::{cmp, fmt, mem, u64};
+use std::{cmp, fmt, u64};
 
 use protobuf::Message as PbMsg;
 use raft::eraftpb::Entry;
 
-use crate::util::{HashMap, HashSet, RAFT_LOG_STATE_KEY};
+use crate::util::{HashMap, RAFT_LOG_STATE_KEY};
 
-use crate::config::Config;
+use crate::config::{Config, RecoveryMode};
 use crate::log_batch::{
     self, Command, CompressionType, LogBatch, LogItemType, OpType, CHECKSUM_LEN, HEADER_LEN,
 };
 use crate::memtable::{EntryIndex, MemTable};
-use crate::metrics::*;
 use crate::pipe_log::{PipeLog, FILE_MAGIC_HEADER, VERSION};
 use crate::{codec, CacheStats, RaftEngine, RaftLocalState, Result};
 
 const SLOTS_COUNT: usize = 128;
-
-#[derive(Clone, Copy, Debug)]
-#[repr(i32)]
-pub enum RecoveryMode {
-    TolerateCorruptedTailRecords = 0,
-    AbsoluteConsistency = 1,
-}
-
-impl From<i32> for RecoveryMode {
-    fn from(i: i32) -> RecoveryMode {
-        assert!(
-            RecoveryMode::TolerateCorruptedTailRecords as i32 <= i
-                && i <= RecoveryMode::AbsoluteConsistency as i32
-        );
-        unsafe { mem::transmute(i) }
-    }
-}
+const REWRITE_ENTRY_COUNT_THRESHOLD: usize = 128;
```

Comment:
I think it is too large. How about 16? Because we assume that we only rewrite these inactive regions, these regions should not have too many entries.